src/ui/map.js

```diff
@@ -3437,10 +3437,21 @@ class Map extends Camera {
 
         const extension = this.painter.context.gl.getExtension('WEBGL_lose_context');
         if (extension) extension.loseContext();
+
+        this._canvas.removeEventListener('webglcontextlost', this._contextLost, false);
+        this._canvas.removeEventListener('webglcontextrestored', this._contextRestored, false);
+
         removeNode(this._canvasContainer);
         removeNode(this._controlContainer);
         removeNode(this._missingCSSCanary);
+
+        this._canvas = (undefined: any);
+        this._canvasContainer = (undefined: any);
+        this._controlContainer = (undefined: any);
+        this._missingCSSCanary = (undefined: any);
```

Comment:
Is it important to remove the DOM element references here if the `map` object itself is discarded? And if so, should we remove more references, in particular those retained transitively — e.g. `map.dragPan._el`, or `map._logoControl._container` etc.?

Reply:
My thought was that suppose the map object is not discared. For example if you do

```js
const container = 
const map = new mapboxgl.Map({
   container,
   ...
});
....
map.remove();
```

After this code, without removing the DOM references then the browser is not able to discard those elements - if you look in Chrome dev tools heap map you'll end up with a bunch of "Discarded HTML*" elements, because the elements have been removed from the DOM but the retaining reference means they can't be garbage collected.

In terms of the other references, for example `map._logoControl` I think is OK because that control is properly removed and the `._controls` array is emptied in the `map.remove()` - so the DOM elements are removed and the controls are discarded.

The `dragPan` I'm not too sure about, I can't see elements/resources being leaked after my fix so I assume those parts are fine.

Reply:
@mourner  I've added a unit test as you described tracking the event listeners added and removed on the container element.

Reply:
`map._logoControl` reference is retained after `map.remove()` even if `_controls` array is emptied. You can see references to the canvas container element in the handler objects too. So I'm just wondering why some of those retained references leak (like those explicitly cleared in this PR) and some don't — maybe that's worth investigating.